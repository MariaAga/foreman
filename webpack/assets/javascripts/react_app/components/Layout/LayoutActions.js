import {
  LAYOUT_INITIALIZE,
  LAYOUT_SHOW_LOADING,
  LAYOUT_HIDE_LOADING,
  LAYOUT_CHANGE_LOCATION,
  LAYOUT_CHANGE_ACTIVE,
  LAYOUT_CHANGE_ORG,
  LAYOUT_CHANGE_IS_NAV_OPEN,
} from './LayoutConstants';

export const initializeLayout = ({
  items,
  activeMenu,
  organization,
  location,
}) => ({
  type: LAYOUT_INITIALIZE,
  payload: {
    items,
    activeMenu,
    organization,
    location,
  },
});

export const showLoading = () => ({
  type: LAYOUT_SHOW_LOADING,
});

export const hideLoading = () => ({
  type: LAYOUT_HIDE_LOADING,
});

export const changeActiveMenu = ({ title }) => ({
  type: LAYOUT_CHANGE_ACTIVE,
  payload: {
    activeMenu: title,
  },
});

export const changeOrganization = organization => ({
  type: LAYOUT_CHANGE_ORG,
  payload: { organization },
});

export const changeLocation = location => ({
  type: LAYOUT_CHANGE_LOCATION,
  payload: {
    location,
  },
});

export const changeIsNavOpen = isNavOpen => ({
  type: LAYOUT_CHANGE_IS_NAV_OPEN,
  payload: {
    isNavOpen,
  },
});
