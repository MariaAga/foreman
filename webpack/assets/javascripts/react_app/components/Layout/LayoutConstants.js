export const LAYOUT_INITIALIZE = 'LAYOUT_INITIALIZE';
export const LAYOUT_SHOW_LOADING = 'LAYOUT_SHOW_LOADING';
export const LAYOUT_HIDE_LOADING = 'LAYOUT_HIDE_LOADING';
export const LAYOUT_CHANGE_IS_NAV_OPEN = 'LAYOUT_CHANGE_IS_NAV_OPEN';
