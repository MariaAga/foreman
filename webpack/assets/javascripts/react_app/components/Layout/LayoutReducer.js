import Immutable from 'seamless-immutable';

import {
  LAYOUT_INITIALIZE,
  LAYOUT_SHOW_LOADING,
  LAYOUT_HIDE_LOADING,
  LAYOUT_CHANGE_IS_NAV_OPEN,
} from './LayoutConstants';

const initialState = Immutable({
  items: [],
  isLoading: false,
  isNavOpen: true,
});

export default (state = initialState, action) => {
  const { payload, type } = action;

  switch (type) {
    case LAYOUT_INITIALIZE:
      return state
        .set('items', payload.items)
        .set('isNavOpen', payload.isNavOpen)
        .set('currentOrganization', payload.organization)
        .set('currentLocation', payload.location);

    case LAYOUT_SHOW_LOADING:
      return state.set('isLoading', true);

    case LAYOUT_HIDE_LOADING:
      return state.set('isLoading', false);

    case LAYOUT_CHANGE_IS_NAV_OPEN:
      return state.set('isNavOpen', payload.isNavOpen);

    default:
      return state;
  }
};
