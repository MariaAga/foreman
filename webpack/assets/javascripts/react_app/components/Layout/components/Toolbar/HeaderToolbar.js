import React from 'react';
import {
  Toolbar,
  ToolbarContent,
  ToolbarGroup,
  ToolbarItem,
} from '@patternfly/react-core';
import TaxonomySwitcher from '../TaxonomySwitcher/TaxonomySwitcher';
import UserDropdowns from './UserDropdowns';
import NotificationContainer from '../../../notifications';
import ImpersonateIcon from '../ImpersonateIcon';
import { layoutPropTypes, layoutDefaultProps } from '../../LayoutHelper';
import './HeaderToolbar.scss';

const HeaderToolbar = ({
  data,
  currentLocation,
  changeLocation,
  currentOrganization,
  changeOrganization,
  isLoading,
  changeActiveMenu,
  ...props
}) => (
  <Toolbar id="data-toolbar" {...props}>
    <ToolbarContent>
      <ToolbarGroup
        className="test-name"
        spaceItems={{ default: 'sm' }}
        visibility={{ default: 'hidden', lg: 'visible' }}
      >
        <TaxonomySwitcher
          taxonomiesBool={data.taxonomies}
          currentLocation={currentLocation}
          locations={
            data.taxonomies.locations ? data.locations.available_locations : []
          }
          onLocationClick={changeLocation}
          currentOrganization={currentOrganization}
          organizations={
            data.taxonomies.organizations
              ? data.orgs.available_organizations
              : []
          }
          onOrgClick={changeOrganization}
          isLoading={isLoading}
        />
      </ToolbarGroup>
      <ToolbarGroup
        alignment={{ default: 'alignRight' }}
        spaceItems={{ default: 'sm' }}
        visibility={{ default: 'hidden', lg: 'visible' }}
      >
        <ToolbarItem>
          <NotificationContainer data={{ url: data.notification_url }} />
        </ToolbarItem>
        {data.user.impersonated_by && (
          <ToolbarItem>
            <ImpersonateIcon
              stopImpersonationUrl={data.stop_impersonation_url}
            />
          </ToolbarItem>
        )}
        <ToolbarItem>
          <UserDropdowns
            notificationUrl={data.notification_url}
            user={data.user}
            changeActiveMenu={changeActiveMenu}
            stopImpersonationUrl={data.stop_impersonation_url}
          />
        </ToolbarItem>
      </ToolbarGroup>
    </ToolbarContent>
  </Toolbar>
);
HeaderToolbar.propTypes = {
  data: layoutPropTypes.data,
  currentLocation: layoutPropTypes.currentLocation,
  changeLocation: layoutPropTypes.changeLocation,
  currentOrganization: layoutPropTypes.currentOrganization,
  changeOrganization: layoutPropTypes.changeOrganization,
  isLoading: layoutPropTypes.isLoading,
  changeActiveMenu: layoutPropTypes.changeActiveMenu,
};
HeaderToolbar.defaultProps = {
  data: layoutDefaultProps.data,
  currentLocation: layoutDefaultProps.currentLocation,
  changeLocation: layoutDefaultProps.changeLocation,
  currentOrganization: layoutDefaultProps.currentOrganization,
  changeOrganization: layoutDefaultProps.changeOrganization,
  isLoading: layoutDefaultProps.isLoading,
  changeActiveMenu: layoutDefaultProps.changeActiveMenu,
};
export default HeaderToolbar;
