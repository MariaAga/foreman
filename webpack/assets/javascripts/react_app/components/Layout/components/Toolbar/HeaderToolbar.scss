#data-toolbar .pf-c-toolbar__content {
  width: 100%;
}
