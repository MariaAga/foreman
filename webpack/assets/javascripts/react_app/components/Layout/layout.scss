@import '~@theforeman/vendor/scss/variables';
@import '../../common/colors.scss';
@import '../../common/variables.scss';

#react-app-root {
  .react-container.container-fluid {
    padding-top: 60px;
    margin-right: auto;
    margin-left: 250px;
  }
}
.collapsed-nav {
  #react-app-root {
    main.pf-c-page__main {
      padding-left: 0;
    }
  }
  #rails-app-content {
    margin-left: 0;
  }
}

#rails-app-content {
  margin-left: 250px;
}
@media only screen and (max-width: $pf-global--breakpoint--md) {
  #react-app-root .react-container,
  #rails-app-content {
    margin-left: 0;
  }
}

#vertical-spinner {
  margin-top: 15px;
  margin-left: 5px;
}

.collapsed-nav {
  @media only screen and (max-width: $pf-global--breakpoint--md) {
    #react-app-root .react-container,
    #rails-app-content {
      margin-left: 0;
    }
  }
}
#page-sidebar {
  position: fixed;
  bottom: 0;
  top: 60px;
  .pf-c-nav__item.hidden-nav-lg {
    display: none;

    @media (max-width: $pf-global--breakpoint--lg) {
      display: block;
    }
  }
}

.pf-c-nav {
  .nav-title-icon {
    pointer-events: none;
    padding-right: 10px;
  }

  .nav-title {
    pointer-events: none;
  }

  .pf-c-nav__item .pf-c-nav__link::before {
    z-index: 1; //to show dividers
  }
}

.pf-c-menu.pf-m-flyout {
  .pf-c-menu__group-title {
    font-size: var(--pf-global--FontSize--lg);
  }
  .pf-c-divider {
    height: 1px;
  }
}
