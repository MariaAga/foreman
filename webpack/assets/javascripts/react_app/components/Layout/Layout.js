import React from 'react';
import {
  Brand,
  Page,
  PageSection,
  PageSectionVariants,
  PageSidebar,
  SkipToContent,
  PageHeader,
} from '@patternfly/react-core';

import { translate as __ } from '../../common/I18n';
import VerticalNav from './components/VerticalNav';
import HeaderToolbar from './components/Toolbar/HeaderToolbar';

import { layoutPropTypes, layoutDefaultProps } from './LayoutHelper';
import './layout.scss';

const Layout = ({
  items,
  data,
  isLoading,
  isNavOpen,
  navigate,
  children,
  changeIsNavOpen,
}) => {
  const onNavToggle = () => {
    isNavOpen ? changeIsNavOpen(false) : changeIsNavOpen(true);
  };
  const header = (
    <PageHeader
      logo={
        <React.Fragment>
          <Brand src={data.logo} alt={data.brand} href={data.root} />
          <span className="navbar-brand-txt">
            <span>{data.brand}</span>
          </span>
        </React.Fragment>
      }
      logoProps={{ href: data.root }}
      showNavToggle
      isManagedSidebar
      onNavToggle={onNavToggle}
      headerTools={<HeaderToolbar {...data} isLoading={isLoading} />}
    />
  );

  const sidebar = (
    <PageSidebar
      isNavOpen={isNavOpen}
      nav={<VerticalNav items={items} navigate={navigate} />}
    />
  );
  const pageId = 'main';
  const pageSkipToContent = (
    <SkipToContent href={`#${pageId}`}>{__('Skip to content')}</SkipToContent>
  );
  return (
    <Page
      header={header}
      sidebar={sidebar}
      skipToContent={pageSkipToContent}
      mainContainerId={pageId}
    >
      <PageSection
        className="react-container nav-pf-persistent-secondary"
        variant={PageSectionVariants.light}
      >
        {children}
      </PageSection>
    </Page>
  );
};
Layout.propTypes = layoutPropTypes;

Layout.defaultProps = layoutDefaultProps;

export default Layout;
