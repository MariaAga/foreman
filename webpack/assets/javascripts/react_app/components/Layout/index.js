import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';

import {
  initializeLayout,
  collapseLayoutMenus,
  expandLayoutMenus,
  changeIsNavOpen,
} from './LayoutActions';
import reducer from './LayoutReducer';
import {
  patternflyMenuItemsSelector,
  selectIsLoading,
  selectIsNavOpen,
} from './LayoutSelectors';
import { combineMenuItems } from './LayoutHelper';

import { getIsNavbarOpen } from './LayoutSessionStorage';

import Layout from './Layout';

const ConnectedLayout = ({ children, data }) => {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(
      initializeLayout({
        items: combineMenuItems(data),
        isNavOpen: getIsNavbarOpen(),
        organization: data.orgs.current_org,
        location: data.locations.current_location,
      })
    );
  }, [data, dispatch]);

  const isNavOpen = useSelector(state => selectIsNavOpen(state));
  useEffect(() => {
    // toggles a class in the body tag, so that the main #rails-app-content container can have the appropriate width
    if (isNavOpen) {
      document.body.classList.add('pf-m-expanded');
    } else {
      document.body.classList.remove('pf-m-expanded');
    }
  }, [isNavOpen]);

  const { push: navigate } = useHistory();
  const items = useSelector(state => patternflyMenuItemsSelector(state));
  const isLoading = useSelector(state => selectIsLoading(state));
  return (
    <Layout
      data={data}
      navigate={navigate}
      items={items}
      isLoading={isLoading}
      isNavOpen={isNavOpen}
      collapseLayoutMenus={() => dispatch(collapseLayoutMenus())}
      expandLayoutMenus={() => dispatch(expandLayoutMenus())}
      changeIsNavOpen={value => dispatch(changeIsNavOpen(value))}
    >
      {children}
    </Layout>
  );
};

// export prop-types
export const { propTypes, defaultProps } = Layout;

ConnectedLayout.propTypes = {
  children: propTypes.children,
  data: propTypes.data,
};

ConnectedLayout.defaultProps = {
  children: defaultProps.children,
  data: defaultProps.data,
};

// export reducers
export const reducers = { layout: reducer };

// export connected component
export default ConnectedLayout;
