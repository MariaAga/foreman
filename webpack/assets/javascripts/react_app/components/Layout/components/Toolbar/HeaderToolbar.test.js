import { testComponentSnapshotsWithFixtures } from '../../../../common/testHelpers';
import { hasTaxonomiesMock } from '../../Layout.fixtures';

import HeaderToolbar from './HeaderToolbar';

const fixtures = {
  'render HeaderToolbar': {
    data: hasTaxonomiesMock.data,
    currentLocation: hasTaxonomiesMock.currentLocation,
    changeLocation: () => {},
    currentOrganization: hasTaxonomiesMock.currentOrganization,
    changeOrganization: () => {},
    isLoading: false,
    changeActiveMenu: () => {},
  },
};

describe('HeaderToolbar', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(HeaderToolbar, fixtures));
});
