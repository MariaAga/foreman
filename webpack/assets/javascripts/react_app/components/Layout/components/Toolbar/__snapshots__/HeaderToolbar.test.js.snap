// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderToolbar rendering render HeaderToolbar 1`] = `
<Toolbar
  id="data-toolbar"
>
  <ToolbarContent
    isExpanded={false}
    showClearFiltersButton={false}
  >
    <ForwardRef
      spaceItems={
        Object {
          "default": "sm",
        }
      }
      visibility={
        Object {
          "default": "hidden",
          "lg": "visible",
        }
      }
    >
      <TaxonomySwitcher
        currentLocation="london"
        currentOrganization="org1"
        isLoading={false}
        locations={
          Array [
            Object {
              "href": "/locations/1-yaml/select",
              "id": 1,
              "title": "yaml",
            },
            Object {
              "href": "/locations/2-london/select",
              "id": 2,
              "title": "london",
            },
            Object {
              "href": "/locations/3-norway/select",
              "id": 3,
              "title": "norway",
            },
          ]
        }
        onLocationClick={[Function]}
        onOrgClick={[Function]}
        organizations={
          Array [
            Object {
              "href": "/organizations/1-org1/select",
              "id": 1,
              "title": "org1",
            },
            Object {
              "href": "/organizations/2-org2/select",
              "id": 2,
              "title": "org2",
            },
          ]
        }
        taxonomiesBool={
          Object {
            "locations": true,
            "organizations": true,
          }
        }
      />
    </ForwardRef>
    <ForwardRef
      alignment={
        Object {
          "default": "alignRight",
        }
      }
      spaceItems={
        Object {
          "default": "sm",
        }
      }
      visibility={
        Object {
          "default": "hidden",
          "lg": "visible",
        }
      }
    >
      <ToolbarItem>
        <Connect(OnClickOutside(notificationContainer))
          data={
            Object {
              "url": "/notification_recipients",
            }
          }
        />
      </ToolbarItem>
      <ToolbarItem>
        <Connect(ImpersonateIcon)
          stopImpersonationUrl="/users/stop_impersonation"
        />
      </ToolbarItem>
      <ToolbarItem>
        <UserDropdowns
          changeActiveMenu={[Function]}
          className=""
          notificationUrl="/notification_recipients"
          stopImpersonationUrl="/users/stop_impersonation"
          user={
            Object {
              "current_user": Object {
                "user": Object {
                  "firstname": "Admin",
                  "id": 4,
                  "lastname": "User",
                  "login": "admin",
                  "name": "Admin User",
                },
              },
              "impersonated_by": true,
              "user_dropdown": Array [
                Object {
                  "children": Array [
                    Object {
                      "href": "/Aa",
                      "isDivider": false,
                      "onClick": [MockFunction],
                      "title": "Aa",
                    },
                    Object {
                      "href": "/Cc",
                      "isDivider": false,
                      "onClick": [MockFunction],
                      "title": "Cc",
                    },
                  ],
                  "icon": "fa fa-user",
                  "name": "User",
                  "type": "sub_menu",
                },
              ],
            }
          }
        />
      </ToolbarItem>
    </ForwardRef>
  </ToolbarContent>
</Toolbar>
`;
